actionpack/lib/action_dispatch/system_testing/test_helpers/screenshot_helper.rb

```diff
@@ -0,0 +1,57 @@
+module ActionDispatch
+  module SystemTesting
+    module TestHelpers
+      # Screenshot helper for system testing
+      module ScreenshotHelper
+        # Takes a screenshot of the current page in the browser.
+        #
+        # +take_screenshot+ can be used at any point in your system tests to take
+        # a screenshot of the current state. This can be useful for debugging or
+        # automating visual testing.
+        def take_screenshot
+          save_image
+          puts "[Screenshot]: #{image_path}"
+          puts display_image
+        end
+
+        # Takes a screenshot of the current page in the browser if the test
+        # failed.
+        #
+        # +take_failed_screenshot+ is included in <tt>system_test_helper.rb</tt> that is
+        # generated with the application. To take screenshots when a test fails
+        # add +take_failed_screenshot+ to the teardown block before clearing
+        # sessions.
+        def take_failed_screenshot
+          take_screenshot unless passed?
+        end
```

Comment:
When a failed screenshot is produced, we should reflect that in the error message somehow. I just got this on a failed test:

```
PostsTest#test_visiting_the_index:
Capybara::ExpectationNotMet: expected to find css "h1" with text "BONKERS" but there were no matches. Also found "Posts", which matched the selector but not all filters.
    test/system/posts_test.rb:7:in `block in <class:PostsTest>'
```

Maybe that could be:

```
PostsTest#test_visiting_the_index:
Capybara::ExpectationNotMet: expected to find css "h1" with text "BONKERS" but there were no matches. Also found "Posts", which matched the selector but not all filters. See a screenshot of the failure in tmp/screenshots/failures_test_visiting_the_index.png.
    test/system/posts_test.rb:7:in `block in <class:PostsTest>'
```